### PR TITLE
APM_Control: Modified PID structure to be a rate PID + FF

### PR DESCRIPTION
**Pull request by ClockeNessMnstr**
The tuning and behavior of the system was confusing.
P was the same as FF so there were two FF's
D was actually a rate P and included the set point.
Old tuning structure is still present but re-mapped.
P->FF, FF_new = old (FF + MAX((gains.P - gains.I * gains.tau) * gains.tau  - gains.D , 0))
D->P (no scaling needed)
new D term is introduced:
rate feedback without setpoint.
low pass filtered. cuttoff currently set at 1 HZ

### CODE REVIEW

**Comment by ClockeNessMnstr** [comment]
I have been looking for history on the "Old PID Laws" referenced in the re-scaling and couldn't find any. 

Has this approach been analyzed having a PID on the rate feedback loop? is that where the scaling of P came from?
P->FF, FF_new = old (FF + MAX((gains.P - gains.I * gains.tau) * gains.tau - gains.D , 0))

File: libraries/APM_Control/AP_PitchController.cpp [hunk 1]
```diff
@@ -165,6 +165,26 @@ int32_t AP_PitchController::_get_rate_out(float desired_rate, float scaler, bool
 	} else {
 		_pid_info.I = 0;
 	}
+	
+	float D_Filt_f3db = 1; // hz for 3 dB cutoff
```
Comment (khancyr) [inline libraries/APM_Control/AP_PitchController.cpp:169]: `static const float D_FILT_F3DB = 1.0f;`

File: libraries/APM_Control/AP_PitchController.cpp [hunk 2]
```diff
@@ -165,6 +165,26 @@ int32_t AP_PitchController::_get_rate_out(float desired_rate, float scaler, bool
...
+	float tau = 1.0f / (2 * 3.14159 * D_Filt_f3db); // tau for filtering
```
Comment (khancyr) [inline libraries/APM_Control/AP_PitchController.cpp:170]: static const float tau = 1.0f / (2.0f * M_PI * D_FILT_F3DB );

File: libraries/APM_Control/AP_PitchController.cpp [hunk 3]
```diff
@@ -175,16 +195,15 @@ int32_t AP_PitchController::_get_rate_out(float desired_rate, float scaler, bool
 	// Calculate equivalent gains so that values for K_P and K_I can be taken across from the old PID law
     // No conversion is required for K_D
     float eas2tas = _ahrs.get_EAS2TAS();
-	float kp_ff = MAX((gains.P - gains.I * gains.tau) * gains.tau  - gains.D , 0) / eas2tas;
+	//float kp_ff = MAX((gains.P - gains.I * gains.tau) * gains.tau  - gains.D , 0) / eas2tas;
```
Comment (khancyr) [inline libraries/APM_Control/AP_PitchController.cpp:198]: if it is unused, then remove it ! We don't keep unused code

**Comment by khancyr** [comment]
hello,

Thanks for contributing. 
Could you made some change on your commit style according to our commit rules (see https://ardupilot.org/dev/docs/submitting-patches-back-to-master.html#submitting-patches-back-to-master)
The commit name should be `APM_Control: Modified PID structure to be a rate PID + FF`
I put some other style fix on the review.

Could you produce some graph or log that show the improvement or what was wrong in the previous code ?

**Comment by ClockeNessMnstr** [comment]
> hello,
> 
> Thanks for contributing.
> Could you made some change on your commit style according to our commit rules (see https://ardupilot.org/dev/docs/submitting-patches-back-to-master.html#submitting-patches-back-to-master)
> The commit name should be `APM_Control: Modified PID structure to be a rate PID + FF`
> I put some other style fix on the review.
> 
> Could you produce some graph or log that show the improvement or what was wrong in the previous code ?

Sure!

**Comment by ClockeNessMnstr** [comment]
I am new to using git. I think I messed this up. I'll update this and bring it back.

**Comment by andyp1per** [comment]
For this to go in you will likely have to prove the value, it would be good to know what you are trying to achieve as its in a complex part of the code. Have you flown with this - are you able to articulate how it is better with this change? Thanks!

**Comment by khancyr** [comment]
@ClockeNessMnstr Check on the wiki, we tried to provide some instruction on how to use git and get the PR nicely done ! Otherwise, ask on gitter/discord for help on git, plenty of people will be able to help you !

**Comment by ClockeNessMnstr** [comment]
> For this to go in you will likely have to prove the value, it would be good to know what you are trying to achieve as its in a complex part of the code. Have you flown with this - are you able to articulate how it is better with this change? Thanks!

I've had success with SITL and plan on flashing my plane soon once I get my repository cleaned up. 

I'm going to get some logs of the tuning process described on the Wiki to compare performance of the process and performance of the resulting controllers. 

The change is functionally just an addition of the derivative on the rate path. So I'm looking to see if that is as effective in flight as it is in the SIM.

**Comment by rmackay9** [comment]
If it were up to me (and it's definitely not) I would use the AC_PID objects because they have a lot of features and are already used in Copter (including TradHeli), Rover and Sub.

**Comment by samuelctabor** [comment]
A related thing that is missing from Plane at present is a way to get the current target roll/pitch, as distinct from the navigation roll/pitch.

Testing our stall detector, which monitors roll error. we found that he navigation roll/pitch can change instantaneously, which cannot be physically achieved, so calculating an angle error based on navigation roll/pitch will always show spikes on changes of waypoint or command, even if the aircraft is perfectly tuned.

In contrast, a target roll/pitch would essentially be the integral of the target rate. and if the target rate is being achieved the corresponding angle error would be zero.

**Comment by ClockeNessMnstr** [comment]
> A related thing that is missing from Plane at present is a way to get the current target roll/pitch, as distinct from the navigation roll/pitch.
> 
> Testing our stall detector, which monitors roll error. we found that he navigation roll/pitch can change instantaneously, which cannot be physically achieved, so calculating an angle error based on navigation roll/pitch will always show spikes on changes of waypoint or command, even if the aircraft is perfectly tuned.
> 
> In contrast, a target roll/pitch would essentially be the integral of the target rate. and if the target rate is being achieved the corresponding angle error would be zero.

Is it actually distinct? I would think, but haven't dug into any NAV controller specifics, that the Target from NAV would be actual target. 

The target isn't perfectly achievable but it's ok for it to change instantly. 
ptch2srv_tconst and rll2srv_tconst define how "aggressively" the controller want's to chase this target. 
maybe you are looking for "(target angle - actual angle) / tconst"? this is the "input filtered" target and goes to the rate controller. 
If not, I've probably mis-understood your statements.

**Comment by samuelctabor** [comment]
> you could filter the setpoint with a low pass filter using rll2srv_tconst. With a well tuned PID loop on the rate, the anglular response should match this.

Correct, this is more or less what we have done, with the addition of the rate limit.

**Comment by ClockeNessMnstr** [comment]
So my little pixhawk1 is essentially hot glued to my motor being crammed into a nano skyhunter. I'm taking out the case and getting some foam for a soft mount setup.
I also have a full size skyhunter to set up if I can't get my vibes down. 

PI controller wasn't even adequately stable with the vibes so I couldn't try a D term. I had to tune softer than my previous feedforward tuning.

**Comment by ClockeNessMnstr** [comment]
Had horrible results testing on my small plane after trying vibration damping. May come back to this when I get a bigger plane running.  moved on to a helicopter project.

**Pull request closed without merging**